lib/src/rules/no_constant_pattern_type_literal.dart

```diff
@@ -0,0 +1,108 @@
+// Copyright (c) 2023, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+
+import '../analyzer.dart';
+
+const _desc = r"Don't use constant patterns with type literals.";
+
+const _details = r'''
+Use `== TypeName` or `TypeName _` instead of type literals in patterns.
+
+**BAD:**
+```dart
+void f(Type x) {
+  if (x case int) {
+    print('int');
+  }
+}
+```
+
+**GOOD:**
+```dart
+void f(Type x) {
+  if (x case == int) {
+    print('int');
+  }
+}
+```
+
+**BAD:**
+```dart
+void f(Object? x) {
+  if (x case int) {
+    print('int');
+  }
+}
+```
+
+**GOOD:**
+```dart
+void f(Object? x) {
+  if (x case int _) {
+    print('int');
+  }
+}
+```
+
+''';
+
+class NoConstantPatternTypeLiteral extends LintRule {
+  static const String lintName = 'no_constant_pattern_type_literal';
```

Comment:
Or `type_literal_in_constant_pattern`?

There's always been some inconsistency between lints names describing what you ought to do and describing what you did wrong. All of the non-lint diagnostics describing what's wrong with the code, and I'm leaning in that direction for this lint, but it's a judgement call.

Reply:
Done.